asset.go

```diff
@@ -0,0 +1,173 @@
+}
+
+// copySymlink copies the symlink, but before doing so it ensures that
+// the symlink is pointing to node inside the asset.
+//
+// Example:
+//
+// Given are local asset $HOME/some/asset and a symlink inside in
+// $HOME/some/asset/some/symlink. If the symlink points to for example
+// to $HOME/foo or to ../../../bar then the symlink is pointing
+// outside the asset and we don't support them.
+func copySymlink(src, dest, imageAssetDir, root string) error {
+	symTarget, err := os.Readlink(src)
+	if err != nil {
+		return err
+	}
+	absolute := filepath.IsAbs(symTarget)
+	if !absolute {
+		symTarget = makeAbsoluteTarget(src, symTarget)
+	}
+	if strings.HasPrefix(symTarget, root) {
+		var err error
+		linkTarget := ""
+		if absolute {
+			linkTarget, err = absoluteSrcTargetToDest(root, symTarget, imageAssetDir)
+		} else {
+			linkTarget, err = filepath.Rel(filepath.Dir(src), symTarget)
+		}
+		if err != nil {
+			return err
+		}
+		if err := os.Symlink(linkTarget, dest); err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("Symlink %s points to %s, which is outside asset %s", src, symTarget, root)
+	}
+	return nil
+}
+
+func copyTree(src, dest, imageAssetDir string) error {
+	return filepath.Walk(src, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		rootLess := path[len(src):]
+		target := filepath.Join(dest, rootLess)
+		mode := info.Mode()
+		switch {
+		case mode.IsDir():
+			err := os.Mkdir(target, mode.Perm())
+			if err != nil {
+				return err
+			}
+		case mode.IsRegular():
+			if err := copyRegularFile(path, target); err != nil {
+				return err
+			}
+		case mode&os.ModeSymlink == os.ModeSymlink:
+			if err := copySymlink(path, target, imageAssetDir, src); err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("Unsupported node (%s) in assets, only regular files, directories and symlinks pointing to node inside asset are supported.", path, mode.String())
+		}
+		return nil
+	})
+}
+
+func replacePlaceholders(path string, paths map[string]string) string {
+	Debug("Processing path: ", path)
+	newPath := path
+	for placeholder, replacement := range paths {
+		newPath = strings.Replace(newPath, placeholder, replacement, -1)
+	}
+	Debug("Processed path: ", newPath)
+	return newPath
+}
+
+func validateAsset(ACIAsset, localAsset string) error {
+	if !filepath.IsAbs(ACIAsset) {
+		return fmt.Errorf("Wrong ACI asset: '%v' - ACI asset has to be absolute path", ACIAsset)
+	}
+	if !filepath.IsAbs(localAsset) {
+		return fmt.Errorf("Wrong local asset: '%v' - local asset has to be absolute path", localAsset)
+	}
+	fi, err := os.Stat(localAsset)
+	if err != nil {
+		return fmt.Errorf("Error stating %v: %v", localAsset, err)
+	}
+	if fi.Mode().IsDir() || fi.Mode().IsRegular() {
+		return nil
+	}
+	return fmt.Errorf("Can't handle local asset %v - not a file, not a dir", fi.Name())
+}
+
+func PrepareAssets(assets []string, rootfs string, paths map[string]string) error {
```

Comment:
No need to export PrepareAssets with a capital P since it is not used outside of package main.